src/middlewared/middlewared/plugins/chart_releases_linux/chart_release.py

```diff
@@ -0,0 +1,223 @@
+
+        catalog = await self.middleware.call(
+            'catalog.query', [['id', '=', data['catalog']]], {'get': True, 'extra': {'item_details': True}}
+        )
+        if data['train'] not in catalog['trains']:
+            raise CallError(f'Unable to locate "{data["train"]}" catalog train.')
+        if data['item'] not in catalog['trains'][data['train']]:
+            raise CallError(f'Unable to locate "{data["item"]}" catalog item.')
+        if data['version'] not in catalog['trains'][data['train']][data['item']]['versions']:
+            raise CallError(f'Unable to locate "{data["version"]}" catalog item version.')
+
+        item_details = catalog['trains'][data['train']][data['item']]['versions'][data['version']]
+        k8s_config = await self.middleware.call('kubernetes.config')
+        release_ds = os.path.join(k8s_config['dataset'], 'releases', data['release_name'])
+        # The idea is to validate the values provided first and if it passes our validation test, we
+        # can move forward with setting up the datasets and installing the catalog item
+        default_values = item_details['values']
+        new_values = copy.deepcopy(default_values)
+        new_values.update(data['values'])
+        new_values, context = await self.normalise_and_validate_values(item_details, new_values, False, release_ds)
+
+        # Now that we have completed validation for the item in question wrt values provided,
+        # we will now perform the following steps
+        # 1) Create release datasets
+        # 2) Copy chart version into release/charts dataset
+        # 3) Install the helm chart
+        # 4) Create storage class
+        storage_class_name = await get_storage_class_name(data['release_name'])
+        try:
+            for dataset in await self.release_datasets(release_ds):
+                if not await self.middleware.call('pool.dataset.query', [['id', '=', dataset]]):
+                    await self.middleware.call('pool.dataset.create', {'name': dataset, 'type': 'FILESYSTEM'})
+
+            chart_path = os.path.join('/mnt', release_ds, 'charts', data['version'])
+            await self.middleware.run_in_thread(lambda: shutil.copytree(item_details['location'], chart_path))
+
+            # Before finally installing the release, we will perform any actions which might be required
+            # for the release to function like creating/deleting ix-volumes
+            await self.perform_actions(context)
+
+            with tempfile.NamedTemporaryFile(mode='w+') as f:
+                f.write(yaml.dump(new_values))
+                f.flush()
+                # We will install the chart now and force the installation in an ix based namespace
+                # https://github.com/helm/helm/issues/5465#issuecomment-473942223
+                cp = await run(
+                    [
+                        'helm', 'install', data['release_name'], chart_path, '-n',
```

Comment:
So this will just deploy the workload in the k8s cluster, the actual pulling of images will happen automatically when k8s accepts pods and schedules them on the node, they will remain in `pending` state until all the required images have been pulled.
I have been thinking of using events for notifications in this case and let install proceed without actually waiting for the pods to come online